src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,572 @@
+        private JsonSeparatedList ParseCommaSeparatedSequenceWorker()
+        {
+            using var _ = ArrayBuilder<JsonNodeOrToken>.GetInstance(out var result);
+            var allProperties = true;
+            while (ShouldConsumeSequenceElement())
+            {
+                var value = ParseValue();
+                allProperties = allProperties && value.Kind == JsonKind.Property;
+                result.Add(value);
+
+                // Try to consume a comma.  If we don't see one, consume an empty one as a placeholder. Create a
+                // diagnostic message depending on if we've seen only properties before this point.  If not, don't give
+                // a message about a missing comma.  Instead, we'll give a specific message that we didn't get a
+                // property when we expected one.
+                if (ShouldConsumeSequenceElement())
+                    result.Add(ConsumeToken(JsonKind.CommaToken, allProperties ? s_commaExpected : null));
+            }
+
+            return new JsonSeparatedList(result.ToImmutable());
+        }
+
+        private bool ShouldConsumeSequenceElement()
+        {
+            if (_currentToken.Kind == JsonKind.EndOfFile)
+                return false;
+
+            if (_currentToken.Kind == JsonKind.CloseBraceToken)
+                return !_inObject;
+
+            if (_currentToken.Kind == JsonKind.CloseBracketToken)
+                return !_inArray;
+
+            if (_currentToken.Kind == JsonKind.CloseParenToken)
+                return !_inConstructor;
+
+            return true;
+        }
+
+        private JsonValueNode ParseValue()
+            => _currentToken.Kind switch
+            {
+                JsonKind.OpenBraceToken => ParseObject(),
+                JsonKind.OpenBracketToken => ParseArray(),
+                JsonKind.CommaToken => ParseCommaValue(),
+                _ => ParseLiteralOrPropertyOrConstructor(),
+            };
+
+        private static void SplitLiteral(JsonToken literalToken, out JsonToken minusToken, out JsonToken newLiteralToken)
+        {
+            minusToken = CreateToken(
+                JsonKind.MinusToken, literalToken.LeadingTrivia,
+                literalToken.VirtualChars.GetSubSequence(new TextSpan(0, 1)),
+                ImmutableArray<JsonTrivia>.Empty);
+            newLiteralToken = CreateToken(
+                literalToken.Kind,
+                ImmutableArray<JsonTrivia>.Empty,
+                literalToken.VirtualChars.GetSubSequence(TextSpan.FromBounds(1, literalToken.VirtualChars.Length)),
+                literalToken.TrailingTrivia,
+                literalToken.Diagnostics);
+        }
+
+        private JsonPropertyNode ParseProperty(JsonToken stringLiteralOrText)
+        {
+            Debug.Assert(_currentToken.Kind == JsonKind.ColonToken);
+            if (stringLiteralOrText.Kind != JsonKind.StringToken)
+                stringLiteralOrText = stringLiteralOrText.With(kind: JsonKind.TextToken);
```

Comment:
Why is this not already the case? What was it before?

Reply:
i'll doc.  but you could write (illegally):  `{ 0: true }`.  so it will be a literal there that we convert to text.